File: rectangular_channel/elm_line.f
      subroutine elm_line
     +
     +  (N
     +  ,ratio
     +  ,X1,Y1
     +  ,X2,Y2
     +  ,sinit
     +  ,Isym
     +  ,Xe,Ye,se
     +  ,Xm,Ym,sm
     +  )

c-----------------------------------------
c FDLIB
c
c Copyright by C. Pozrikidis, 1999
c All rights reserved.
c
c This program is to be used only under the
c stipulations of the licensing agreement.
c----------------------------------------

c-----------------------------------------------------
c Disretization of a line segment into N elements
c
c  X1,Y1: coordinates of the first point
c  X2,Y2: coordinates of the last point
c
c  ratio: 
c
c     If Isym = 0, ratio of length of LAST to FIRST element
c     If Isym = 1, ratio of length of MID  to FIRST element
c
c  alpha: geometric factor ratio
c
c  sinit: specified arc length at (X1, Y1)
c
c  se: arc length at the element end-nodes
c  sm: arc length at the element mid-nodes
c
c  xe,ye: end nodes
c  xm,ym: mid nodes
c
c-----------------------------------------------------

      Implicit Double Precision (a-h,o-z)

      Dimension Xe(100),Ye(100),se(100)
      Dimension Xm(100),Ym(100),sm(100)

c------------
c one element
c------------

      If(N.eq.1) then

       xe(1) = X1 
       ye(1) = Y1 

       xe(2) = X2 
       ye(2) = Y2 

       se(1) = sinit
       se(2) = se(1)+Dsqrt( (X2-X1)**2+(Y2-Y1)**2)

       Go to 99

      End If

c--------
c prepare
c--------

      N1 = N+1

c--------------------
c biased distribution
c--------------------

      If(Isym.eq.0) then

      If(ratio.eq.1.000) then
        alpha  = 1.0D0
        factor = 1.0D0/N
      Else
        texp   = 1.0D0/(N-1.0D0)
        alpha  = ratio**texp
        factor = (1.0D0-alpha)/(1.0D0-alpha**N)
      End If

      deltax = (x2-x1) * factor   ! x length of first element
      deltay = (y2-y1) * factor   ! y length of first element 

      Xe(1) = X1     ! first point
      Ye(1) = Y1
      se(1) = sinit

      Do i=2,N1
        Xe(i)  = Xe(i-1)+deltax
        Ye(i)  = Ye(i-1)+deltaY
        se(i)  = se(i-1)+Dsqrt(deltax**2+deltay**2)
        deltax = deltax*alpha
        deltaY = deltay*alpha
      End Do

      Go to 99

      End If

c----------------------
c symmetric distribution
c even number of points
c----------------------

      If(mod(N,2).eq.0) then
   
      Xh = 0.5D0*(X1+X2)      ! mid-point
      Yh = 0.5D0*(Y1+Y2)

      Nh  = N/2
      Nh1 = Nh+1

      If(ratio.eq.1.000) then
        alpha  = 1.0D0
        factor = 1.0D0/Nh
      Else
        texp   = 1.0D0/(Nh-1.0)
        alpha  = ratio**texp
        factor = (1.0D0-alpha)/(1.0D0-alpha**Nh)
      End If

      deltax = (xh-x1) * factor   ! x length of first element
      deltay = (yh-y1) * factor   ! y length of first element 

      Xe(1) = X1    ! first point
      Ye(1) = Y1
      se(1) = sinit

      Do i=2,Nh1
        Xe(i)  = Xe(i-1)+deltax
        Ye(i)  = Ye(i-1)+deltaY
        se(i)  = se(i-1)+sqrt(deltax**2+deltay**2)
        deltax = deltax*alpha
        deltaY = deltay*alpha
      End Do

      deltax = deltax/alpha
      deltaY = deltay/alpha

      Do i=Nh1+1,N1
        Xe(i)  = Xe(i-1)+deltax
        Ye(i)  = Ye(i-1)+deltaY
        se(i)  = se(i-1)+sqrt(deltax**2+deltay**2)
        deltax = deltax/alpha
        deltay = deltay/alpha
      End Do

      Go to 99

      End If

c-----------------------
c symmetric distribution
c odd number of points
c-----------------------

      If(ratio.eq.1.000) then
        alpha  = 1.0D0
        factor = 1.0D0/N1
      Else
        texp   = 2.0D0/(N-1.0D0)
        alpha  = ratio**texp
        tmp1   = 0.50D0*(N+1.0D0)
        tmp2   = 0.50D0*(N-1.0D0)
        factor = (1.0D0-alpha)/(2.0D0-alpha**tmp1-alpha**tmp2)
      End If

      deltax = (x2-x1) * factor   ! x length of first element
      deltay = (y2-y1) * factor   ! y length of first element 

      Xe(1) = X1     ! first point
      Ye(1) = Y1
      se(1) = sinit

      Do i=2,(N+3)/2
        Xe(i)  = Xe(i-1)+deltax
        Ye(i)  = Ye(i-1)+deltaY
        se(i)  = se(i-1)+sqrt(deltax**2+deltay**2)
        deltax = deltax*alpha
        deltaY = deltay*alpha
      End Do

      deltax = deltax/(alpha**2)
      deltay = deltay/(alpha**2)

      Do i=(N+5)/2,N1
        Xe(i)  = Xe(i-1)+deltax
        Ye(i)  = Ye(i-1)+deltaY
        se(i)  = se(i-1)+sqrt(deltax**2+deltay**2)
        deltax = deltax/alpha
        deltaY = deltay/alpha
      End Do

c-----
c Done
c-----

  99  Continue

c-------------------
c compute mid-points
c-------------------

      Do i=1,N
       xm(i) = 0.5D0*(xe(i)+xe(i+1))
       ym(i) = 0.5D0*(ye(i)+ye(i+1))
       sm(i) = 0.5D0*(se(i)+se(i+1))
      End Do

c-----
c Done
c-----

      write (6,*) "elm_line: Geometric ratio: ",alpha

      Return
      End

File: rectangular_channel/elm_line.m
function [xe,ye,se,xm,ym,sm]...
 ...
      = elm_line (N,ratio,x1,y1,x2,y2,sinit,Isym)

%-----------------------------------------
% FDLIB, BEMLIB
%
% Copyright by C. Pozrikidis, 1999
% All rights reserved.
%
% This program is to be used only under the
% stipulations of the licensing agreement.
%----------------------------------------

%------------------------------------------------
% Disretization of a line segment in the xy plane
% into N elements
%
% x1,y1: coordinates of the first point
% x2,y2: coordinates of the last point
%
% ratio: 
% if Isym = 0: ratio of length of LAST to FIRST element
% if Isym = 1: ratio of length of MID  to FIRST element
%
% alpha: geometric factor ratio
%
% sinit: specified arc length at (x1, y1)
%
% se: arc length at the element end-nodes
% sm: arc length at the element mid-nodes
%
% xe,ye: end nodes
% xm,ym: mid nodes
%-----------------------------------------------------

%------------
% one element
%------------

if(N==1) 

   xe(1) = x1;
   ye(1) = y1;
   xe(2) = x2; 
   ye(2) = y2;
   se(1) = sinit;
   se(2) = se(1)+sqrt( (x2-x1)^2+(y2-y1)^2);

%--------------------
% biased distribution
%--------------------

elseif(Isym==0)

  if(ratio==1.000)
    alpha  = 1.0;
    factor = 1.0/N;
  else
    texp   = 1.0/(N-1.0);
    alpha  = ratio^texp;
    factor = (1.0-alpha)/(1.0-alpha^N);
  end

  deltax = (x2-x1)*factor;  % x length of first element
  deltay = (y2-y1)*factor;  % y length of first element 

  xe(1) = x1;    % first point
  ye(1) = y1;
  se(1) = sinit;

  for i=2:N+1
    xe(i)  = xe(i-1)+deltax;
    ye(i)  = ye(i-1)+deltay;
    se(i)  = se(i-1)+sqrt(deltax^2+deltay^2);
    deltax = deltax*alpha;
    deltay = deltay*alpha;
  end

%----------------------
% symmetric distribution
%----------------------

elseif(Isym==1)

 if(N==2)

   xe(1) = x1;
   ye(1) = y1;

   xe(2) = 0.5*(x1+x2);
   ye(2) = 0.5*(y1+y2);

   xe(3) = x2;
   ye(3) = y2;

   se(1) = sinit;
   se(2) = se(1)+sqrt( (x(2)-x1)^2+(y(2)-y1)^2);
   se(3) = se(2)+sqrt( (x2-x(2))^2+(y2-y(2))^2);

 elseif(mod(N,2)==0) % even number of points

   xh = 0.5*(x1+x2);     % mid-point
   yh = 0.5*(y1+y2);
   
   Nh  = N/2;
   Nh1 = Nh+1;

   if(ratio==1.000)
        alpha  = 1.0;
        factor = 1.0/Nh;
   else
        texp   = 1.0/(Nh-1.0);
        alpha  = ratio^texp;
        factor = (1.0-alpha)/(1.0-alpha^Nh);
   end

   deltax = (xh-x1)*factor;  % x length of first element
   deltay = (yh-y1)*factor;  % y length of first element 

   xe(1) = x1;   % first point
   ye(1) = y1;
   se(1) = sinit;

   for i=2:Nh1
        xe(i)  = xe(i-1)+deltax;
        ye(i)  = ye(i-1)+deltay;
        se(i)  = se(i-1)+sqrt(deltax^2+deltay^2);
        deltax = deltax*alpha;
        deltay = deltay*alpha;
   end

   deltax = deltax/alpha;
   deltay = deltay/alpha;

   for i=Nh1+1:N+1
     xe(i)  = xe(i-1)+deltax;
     ye(i)  = ye(i-1)+deltay;
     se(i)  = se(i-1)+sqrt(deltax^2+deltay^2);
     deltax = deltax/alpha;
     deltay = deltay/alpha;
   end

 else

% odd number of points

     if(ratio==1.000)
        alpha  = 1.0;
        factor = 1.0/(N+1);
      else
        texp   = 2.0/(N-1.0);
        alpha  = ratio^texp;
        tmp1   = 0.50*(N+1.0);
        tmp2   = 0.50*(N-1.0);
        factor = (1.0-alpha)/(2.0-alpha^tmp1-alpha^tmp2);
     end

     deltax = (x2-x1)*factor;   % x length of first element
     deltay = (y2-y1)*factor;  % y length of first element 

     xe(1) = x1;    % first point
     ye(1) = y1;
     se(1) = sinit

     for i=2:(N+3)/2
        xe(i)  = xe(i-1)+deltax;
        ye(i)  = ye(i-1)+deltay;
        se(i)  = se(i-1)+sqrt(deltax^2+deltay^2);
        deltax = deltax*alpha;
        deltay = deltay*alpha;
     end

     deltax = deltax/(alpha^2);
     deltay = deltay/(alpha^2);

     for i=(N+5)/2:N+1
        xe(i)  = xe(i-1)+deltax;
        ye(i)  = ye(i-1)+deltay;
        se(i)  = se(i-1)+sqrt(deltax^2+deltay^2);
        deltax = deltax/alpha;
        deltay = deltay/alpha;
     end

end

%---
end
%---

%-------------------
% compute mid-points
%-------------------

  for i=1:N
    xm(i) = 0.5*(xe(i)+xe(i+1));
    ym(i) = 0.5*(ye(i)+ye(i+1));
    sm(i) = 0.5*(se(i)+se(i+1));
  end

%-----
% done
%-----

return
